### PR TITLE
Access Groups, Dm Messages, Group Chats

### PR DESCRIPTION
**Pull request by AeonSw4n**

### CODE REVIEW

File: lib/blockchain.go [hunk 1]
```diff
@@ -2017,6 +2017,17 @@ func (bc *Blockchain) ProcessBlock(desoBlock *MsgDeSoBlock, verifySignatures boo
 		if err = bc.postgres.UpsertBlock(nodeToValidate); err != nil {
 			err = errors.Wrapf(err, "ProcessBlock: Problem saving block with StatusBlockStored")
 		}
+
+		// We're not storing blocks in postgres, so we should store the actual blocks in the badgerdb.
```
Comment (lazynina) [inline lib/blockchain.go:2021]: Well it does look like we are storing blocks in the upsert above. is the issue that we're just not checking postgres in GetBlock?

File: lib/blockchain.go [hunk 2]
```diff
@@ -2617,23 +2630,39 @@ func (bc *Blockchain) DisconnectBlocksToHeight(blockHeight uint64) error {
 			}
 
 			// Set the best node hash to the new tip.
-			if err := PutBestHashWithTxn(txn, nil, &prevHash, ChainTypeDeSoBlock); err != nil {
-				return err
+			if bc.postgres != nil {
+				if err := bc.postgres.UpsertChain(MAIN_CHAIN, &prevHash); err != nil {
```
Comment (lazynina) [inline lib/blockchain.go:2634]: it would be better if we could move this postgres v badger logic to the db adapter

File: lib/blockchain.go [hunk 3]
```diff
@@ -2617,23 +2630,39 @@ func (bc *Blockchain) DisconnectBlocksToHeight(blockHeight uint64) error {
...
-			if err := DeleteBlockRewardWithTxn(txn, nil, blockToDetach); err != nil {
+			if err := DeleteBlockRewardWithTxn(txn, snap, blockToDetach); err != nil {
 				return errors.Wrapf(err, "DisconnectBlocksToHeight: Problem deleting block reward")
 			}
 
+			// Revert the detached block's status to StatusHeaderValidated and save the blockNode to the db.
 			node.Status = StatusHeaderValidated
-			if err := PutHeightHashToNodeInfoWithTxn(txn, nil, node, false); err != nil {
-				return errors.Wrapf(err, "DisconnectBlocksToHeight: Problem deleting height hash to node info")
+			if bc.postgres != nil {
```
Comment (lazynina) [inline lib/blockchain.go:2655]: would prefer to move pg v. badger logic to db adapter.

File: lib/db_adapter.go [hunk 4]
```diff
@@ -131,3 +132,76 @@ func (adapter *DbAdapter) GetPKIDForPublicKey(pkBytes []byte) *PKID {
...
 }
+
+//
+// AccessGroups
+//
+
+func (adapter *DbAdapter) GetAccessGroupEntryByAccessGroupId(accessGroupId *AccessGroupId) (*AccessGroupEntry, error) {
+	if accessGroupId == nil {
+		glog.Errorf("GetAccessGroupEntryByAccessGroupId: Called with nil accessGroupId, this should never happen")
+		return nil, nil
```
Comment (lazynina) [inline lib/db_adapter.go:143]: why glog it when this function returns an error?

File: lib/db_adapter.go [hunk 5]
```diff
@@ -131,3 +132,76 @@ func (adapter *DbAdapter) GetPKIDForPublicKey(pkBytes []byte) *PKID {
...
+	}
+
+	if adapter.postgresDb != nil {
+		pgAccessGroup := adapter.postgresDb.GetAccessGroupByAccessGroupId(accessGroupId)
+		if pgAccessGroup == nil {
+			return nil, nil
+		}
+		return pgAccessGroup.ToAccessGroupEntry(), nil
```
Comment (lazynina) [inline lib/db_adapter.go:151]: you can update `ToAccessGroupEntry` to check if the pgAccessGroup is nil and we can simplify this block of code to be
`return pgAccessGroup.ToAccessGroupEntry()`

File: lib/mempool.go [hunk 6]
```diff
@@ -1787,6 +1787,7 @@ func ComputeTransactionMetadata(txn *MsgDeSoTxn, utxoView *UtxoView, blockHash *
 		}
 
 	}
+	// TODO: Do we need something for access groups and new dm/group chats here?
```
Comment (lazynina) [inline lib/mempool.go:1790]: yes please

**Review by lazynina — COMMENTED** [review]
Checking in the first chunk of my review. Basically looked through the block_view files (need to dig into tests a little bit more), but wanted to get you the feedback I have so far.

Still need to look into
1. db_utils
2. postgres
3. network
4. server

File: lib/block_view_flush.go [hunk 7]
```diff
@@ -924,6 +930,8 @@ func (bav *UtxoView) _flushDAOCoinBalanceEntriesToDbWithTxn(txn *badger.Txn, blo
 
 func (bav *UtxoView) _flushDerivedKeyEntryToDbWithTxn(txn *badger.Txn, blockHeight uint64) error {
 	glog.V(2).Infof("_flushDerivedKeyEntryToDbWithTxn: flushing %d mappings", len(bav.DerivedKeyToDerivedEntry))
+	numDeleted := 0
+	numPut := 0
```
[the remark this answers is not in the document]
Comment (AeonSw4n) [reply lib/block_view_flush.go:934]: Cause otherwise they get reset to 0 everytime lol

File: lib/db_utils.go [hunk 8]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+}
+
+func DBGetAccessGroupIdsForOwner(db *badger.DB, snap *Snapshot, accessGroupOwnerPublicKey PublicKey) ([]*AccessGroupId, error) {
+	var err error
+	var ret []*AccessGroupId
+	err = db.View(func(txn *badger.Txn) error {
+		ret, err = DBGetAccessGroupIdsForOwnerWithTxn(txn, snap, accessGroupOwnerPublicKey)
+		return err
+	})
+	if err != nil && err != badger.ErrKeyNotFound {
```
Comment (lazynina) [inline lib/db_utils.go:1852]: when would we receive a ErrKeyNotFound here?

File: lib/db_utils.go [hunk 9]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+	}
+	return ret, nil
+}
+
+// DBGetAccessGroupExistenceByAccessGroupIdWithTxn checks whether an access group exists in the db using some
+// optimized db queries.
+func DBGetAccessGroupExistenceByAccessGroupIdWithTxn(txn *badger.Txn, snap *Snapshot,
+	accessGroupOwnerPublicKey *PublicKey, accessGroupKeyName *GroupKeyName) (bool, error) {
+
+	prefix := _dbKeyForAccessGroupEntry(*accessGroupOwnerPublicKey, *accessGroupKeyName)
```
Comment (lazynina) [inline lib/db_utils.go:1822]: what happens if we're passed nil for either param?

File: lib/db_utils.go [hunk 10]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+		return nil, errors.Wrapf(err, "DBGetAccessGroupEntryByAccessGroupId: Problem getting access group entry")
+	}
+
+	return ret, nil
+}
+
+func DBGetAccessGroupEntryByAccessGroupIdWithTxn(txn *badger.Txn, snap *Snapshot,
+	accessGroupOwnerPublicKey *PublicKey, accessGroupKeyName *GroupKeyName) (*AccessGroupEntry, error) {
+
+	prefix := _dbKeyForAccessGroupEntry(*accessGroupOwnerPublicKey, *accessGroupKeyName)
```
Comment (lazynina) [inline lib/db_utils.go:1785]: what happens if we're passed nil for either param?

File: lib/db_utils.go [hunk 11]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+		return DBPutAccessGroupEntryWithTxn(txn, snap, blockHeight, accessGroupEntry)
+	})
+	if err != nil {
+		return errors.Wrapf(err, "DBPutAccessGroupEntry: Problem putting access group entry")
+	}
+	return nil
+}
+
+func DBPutAccessGroupEntryWithTxn(txn *badger.Txn, snap *Snapshot, blockHeight uint64, accessGroupEntry *AccessGroupEntry) error {
+	prefix := _dbKeyForAccessGroupEntry(*accessGroupEntry.AccessGroupOwnerPublicKey, *accessGroupEntry.AccessGroupKeyName)
```
Comment (lazynina) [inline lib/db_utils.go:1894]: what happens if we're given a nil accessGroupEntry?

File: lib/db_utils.go [hunk 12]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+	accessGroupMemberPublicKey PublicKey, accessGroupOwnerPublicKey PublicKey, accessGroupKeyName GroupKeyName) []byte {
+
+	prefixCopy := append([]byte{}, Prefixes.PrefixAccessGroupMembershipIndex...)
+	prefixCopy = append(prefixCopy, accessGroupMemberPublicKey.ToBytes()...)
+	prefixCopy = append(prefixCopy, accessGroupOwnerPublicKey.ToBytes()...)
+	prefixCopy = append(prefixCopy, accessGroupKeyName.ToBytes()...)
+	return prefixCopy
+}
+
+func _dbSeekPrefixForAccessGroupMemberEntry(accessGroupMemberPublicKey PublicKey) []byte {
```
Comment (lazynina) [inline lib/db_utils.go:1942]: might we want a seek function for accessGroupMember + groupOwner? "what are all my group chat with User X" use case. not critical now, but food for thought.

File: lib/db_utils.go [hunk 13]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+
+func DBPutAccessGroupMemberEntryWithTxn(txn *badger.Txn, snap *Snapshot, blockHeight uint64,
+	accessGroupMemberEntry *AccessGroupMemberEntry, accessGroupOwnerPublicKey PublicKey, groupKeyName GroupKeyName) error {
+
+	if accessGroupMemberEntry == nil || accessGroupMemberEntry.AccessGroupMemberPublicKey == nil {
+		return fmt.Errorf("DBPutAccessGroupMemberEntryWithTxn: accessGroupMemberEntry is nil or " +
+			"accessGroupMemberEntry.AccessGroupMemberPublicKey is nil")
+	}
+	if reflect.DeepEqual(groupKeyName.ToBytes(), BaseGroupKeyName().ToBytes()) {
+		glog.Errorf("DBPutAccessGroupMemberEntryWithTxn: groupKeyName is empty")
```
Comment (lazynina) [inline lib/db_utils.go:2067]: what do we want to do here? do we save base key? if so, I would remove the if statement. reflect.DeepEqual is expensive relatively.

File: lib/db_utils.go [hunk 14]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+	}
+
+	prefix := _dbKeyForAccessGroupMemberEntry(
```
Comment (lazynina) [inline lib/db_utils.go:2070]: Isn't this really the key, not a prefix?

File: lib/db_utils.go [hunk 15]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+	if err != nil {
+		return errors.Wrapf(err, "DBDeleteAccessGroupMemberEntry: Problem deleting access group member entry")
+	}
+	return nil
+}
+
+func DBDeleteAccessGroupMemberEntryWithTxn(txn *badger.Txn, snap *Snapshot,
+	accessGroupMemberPublicKey PublicKey, accessGroupOwnerPublicKey PublicKey, accessGroupKeyName GroupKeyName) error {
+
+	prefix := _dbKeyForAccessGroupMemberEntry(accessGroupMemberPublicKey, accessGroupOwnerPublicKey, accessGroupKeyName)
```
Comment (lazynina) [inline lib/db_utils.go:2094]: isn't this a key, not a prefix?

File: lib/db_utils.go [hunk 16]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+	groupOwnerPublicKey PublicKey, groupKeyName GroupKeyName, startingAccessGroupMemberPublicKeyBytes []byte,
+	maxMembersToFetch uint32) (_accessGroupMemberPublicKeys []*PublicKey, _err error) {
+
+	var accessGroupMembers []*PublicKey
+	prefix := _dbSeekPrefixForAccessGroupMemberEnumerationIndex(groupOwnerPublicKey, groupKeyName)
+	startKey := append(prefix, startingAccessGroupMemberPublicKeyBytes...)
+
+	keysFound := _enumeratePaginatedLimitedKeysForPrefixWithTxn(txn, prefix, startKey, maxMembersToFetch)
+	// Sanity-check that we don't return the starting key.
+	if len(keysFound) > 0 && bytes.Compare(startKey, keysFound[0]) == 0 {
```
Comment (lazynina) [inline lib/db_utils.go:2194]: This seems a little hacky

File: lib/db_utils.go [hunk 17]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+		// We will fetch one more key after the last key found and append it to the list of keys found, with the first element removed.
+		additionalKey := _enumeratePaginatedLimitedKeysForPrefixWithTxn(txn, prefix, keysFound[len(keysFound)-1], 2)
+		keysFound = append(keysFound[1:], additionalKey[1:]...)
+	}
+
+	for _, key := range keysFound {
+		memberPublicKeyBytes := key[len(prefix):]
```
Comment (lazynina) [inline lib/db_utils.go:2201]: we must be passing in the empty public key to get the first page, right?

File: lib/db_utils.go [hunk 18]
```diff
@@ -1643,6 +1731,546 @@ func DBGetAllUserGroupEntries(handle *badger.DB, ownerPublicKey []byte) ([]*Mess
...
+		accessGroupMembers = append(accessGroupMembers, NewPublicKey(memberPublicKeyBytes))
+	}
+
+	// Sanity-check that the public keys we found are actually lexicographically sorted.
+	for ii := 1; ii < len(accessGroupMembers); ii++ {
```
Comment (lazynina) [inline lib/db_utils.go:2206]: this seems unnecessary since badger is sorted and these compares could be costly.

**Review by lazynina — CHANGES_REQUESTED** [review]

File: lib/block_view_access_group.go [hunk 19]
```diff
@@ -0,0 +1,534 @@
...
+	if groupOwnerPublicKey == nil || groupKeyName == nil {
+		return nil, fmt.Errorf("GetAccessGroupEntry: Called with nil parameter(s)")
+	}
+
+	accessGroupKey := NewAccessGroupId(groupOwnerPublicKey, groupKeyName[:])
+
+	// If the group has already been fetched in this utxoView, then we get it directly from there.
+	if mapValue, exists := bav.AccessGroupIdToAccessGroupEntry[*accessGroupKey]; exists {
+		return mapValue, nil
+	}
```
[the remark this answers is not in the document]
Comment (AeonSw4n) [reply lib/block_view_access_group.go:26]: I think that was just some old code, fixed now.

**Review by AeonSw4n — COMMENTED** [review]
First pass on @lazynina review

**Review by diamondhands0 — APPROVED** [review]

**Pull request merged**